### ISSUE TITLE
What if not all actions are valid

### ISSUE THREAD

**Issue by imandr** [keras-rl/keras-rl#211]
I have a question. How do you suggest to handle the case when not every action is allowed in every state. For example, if I am modeling a lunar lander then my action would be to fire or not to fire the engine. But when the lander runs out of fuel, it can not fire the engine any more, so one of the actions is not allowed.

Thanks!

**Comment by mattdornfeld** [comment]
One thing I’ve been trying is setting up my Environment so that he agent ends up doing nothing if it tries an illegal action. I’m sure there’s a more sophisticated way of handling this but it’s something I’ve been playing with.

**Comment by imandr** [comment]
I tried that, but I am not sure that is correct way to handle that in the DQN case because when the discounted reward is calculated as

   R = r + gamma*max(q_next_step)

and q_next_step does not take into account that the "best" action may turn out to be invalid, and in fact will be a no-op, that makes the q_next_step incorrect

**Comment by mattdornfeld** [comment]
Well it wouldn’t be a no op. If the lander tried to fire the engine when there’s no fuel you would just convert that action to no fire in the environment before applying it. The idea being that the agent would learn that trying to fire the engine when there’s no fuel is equivalent to not firing.

**Comment by imandr** [comment]
I see your point.

Ok, what if there is no obvious or logical way to convert an invalid action into a valid one. For example, if I am modeling a game like tic-tac-toe ?

**Comment by mattdornfeld** [comment]
Yeah than that approach wouldn’t work. They’re has to be an idempotent
action.

**Comment by bklebel** [comment]
I usually try to teach an agent to only take valid actions, by giving negative rewards and ending the episode in case of the agent taking an invalid action. For example in the case of tic-tac-toe, if you try to mark a spot which is already marked, you loose the game immediately. 

In case of the lunar lander, I would think that if your fuel runs out before touchdown, you are going to crash anyways, right? Nevertheless, as was stated above, if there is no fuel, whatever you want to do, the engine will not fire, so the actions are equivalent. Still, I think the question is what you want an agent to learn exactly. Should he still try to fire without any fuel, because if he had fuel, that would be the right choice? Or should he learn to use his fuel more efficiently, so he does not run out of it? 
Even more, does your agent know how much fuel he has left?

**Comment by campbelljc** [comment]
I ran into this same issue a while ago. In my environment there was a large number of functionally impossible actions as @imandr described. To deal with it, I take the max in the backward step over only the q-values corresponding to the possible actions, as well as taking the max/etc in the policy action selection over only the possible actions.

**Comment by djaney** [comment]
I also ran into this issue. What I tried to do is to make the model have 2 inputs 1 is the actual input and the other is a mask of valid actions so it should only predict valid actions. I'm still half way though, there are internal problems like invalid actions are still being selected during exploration.

**Comment by dickreuter** [comment]
I'm facing a similar issue with teaching a Deep Q agent to learn to play poker. I've created an environment here:
https://github.com/dickreuter/neuron_poker

During training, the agent often takes illegal actions. I'm giving it a strong negative reward in that case and have it retry again. But in many cases the agent keeps on performing the same action. Any suggestions how to best address this would be appreciated.

**Comment by kk-55** [comment]
@djaney: I think you can use your mask of valid actions during exploration, too. Just ignore the invalid actions during exploration.

@dickreuter: I think giving the agent a large negative reward still leads to executing illegal actions. As @djaney mentioned, use a mask indicating which actions are legal/illegal. When using softmax activation in your output layer, you can "zeroing out" the probability of illegal actions (set these action logits to "minus infinity").
